### PR TITLE
Close Loop - VMWare Bug (1679225) and Discovery Bug (1367549)

### PR DESCRIPTION
**Pull request by jyejare**
## Automated:

- VMWare Bug (1679225)
- Discovery Bug (1367549)

### CODE REVIEW

**Review by rplevka — CHANGES_REQUESTED** [review]
I'd reconsider the need of manipulating of the backend `dhcp` service.
All you have to do is to define a new `Subnet` entity with the subset of the range.
The algorithm to fetch an unused ip will be restricted to this.

**Comment by JacobCallahan** [comment]
@jyejare  can you revisit this PR when you get a chance? Thanks!

**Comment by omkarkhatavkar** [comment]
Can one of the admins verify this patch?

File: tests/foreman/api/test_discoveredhost.py [hunk 1]
```diff
@@ -288,6 +288,85 @@ def test_positive_provision_pxe_host(self):
                 query={'search': 'name={}'.format(discovered_host.name)}
             )
 
+    @tier3
+    def test_positive_provision_pxe_host_dhcp_change(self):
+        """Provision a pxe-based discovered hosts
```
[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py]: OOPs the copy paste error :)

File: tests/foreman/api/test_discoveredhost.py [hunk 2]
```diff
@@ -288,6 +288,85 @@ def test_positive_provision_pxe_host(self):
...
+                    query={
+                        'search': 'name={}.{}'.format(
+                            discovered_host.name, self.configured_env['domain']['name']
+                        )
+                    }
+                )[0]
+                assert int(provisioned_host.ip.split('.')[-1]) >= int(
+                    new_subnet_from.split('.')[-1]
+                )
+                assert int(provisioned_host.ip.split('.')[-1]) <= int(old_sub_to_4o)
```
[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py]: @rplevka
1. Changing the range from subnet entity doesnt changes the range in dhcpd.conf
2. Discovered hosts during discovery gets the IP from Satellite server DHCP that dhcpd.conf
3. But While provisioning that host,  the IP is assigned from subnet entity range.

And thats we have to check here for the bug - 1367549.

File: tests/foreman/api/test_discoveredhost.py [hunk 3]
```diff
@@ -288,6 +288,85 @@ def test_positive_provision_pxe_host(self):
...
+        subnet = entities.Subnet(id=self.configured_env['subnet']['id']).read()
+        # Updating satellite subnet component and dhcp conf ranges
+        # Storing now for restoring later
+        old_sub_from = subnet.from_
+        old_sub_to = subnet.to
+        old_sub_from_4o = old_sub_from.split('.')[-1]
+        old_sub_to_4o = old_sub_to.split('.')[-1]
+        # Calculating Subnet's new `from` range in Satellite Subnet Component
+        new_subnet_from = subnet.from_.replace(
+            f'.{old_sub_from_4o}', f'.{str(int(old_sub_to_4o)-9)}'
```
[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py]: No worries, at least the static hosts(qe-sat69-rhel-tier3) right now we are using to run these tests has a big pool of IPs to not to fail this test ! :)

File: tests/foreman/api/test_discoveredhost.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py]: @rplevka Also these are the official steps to verify that bug from DEVs :)

File: tests/foreman/api/test_discoveredhost.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py]: @rplevka Do you have any alternate for that btw ? I know we can find the current range and find the middle of it, but feels unnecessary complicating things.

[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py]: Modified the code to not to fail that way !

**Comment by jyejare** [comment]
@rplevka Answered and fixed your comments. So please revisit this long pending PR @rplevka @JacobCallahan and lets get this in.

Sorry for taking long time to revisit.

File: tests/foreman/api/test_discoveredhost.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rplevka) [reply tests/foreman/api/test_discoveredhost.py:411]: @jyejare that's what i'm talking about -
All you have to do is to provision a host using a default `dhcpd.conf` config,
then simply update the Subnet entity attributes to narrow the old range down (not touching the dhcpd.conf) and turn that discovered host into a managed Host using a satellite.
It should reflect the ip settings from the subnet entity (the ip of the created host should be different from the one assigned to discovered host by dhcpd)

File: tests/foreman/api/test_discoveredhost.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rplevka) [reply tests/foreman/api/test_discoveredhost.py]: @jyejare defining an `ipaddress` module object and manipulating it is actually way simpler,
but what I'd do would be to take the very last usable host ip from the old range and make the new range to only allow this single IP.
```
import ipaddress

hosts = [ h for h in ipaddress.ip_network('192.168.1.0/24')][1:-1]
new_range_from = new_range_to = hosts[:-1]
```

..something among those lines.

File: tests/foreman/api/test_discoveredhost.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py:411]: @rplevka Yeah ! Thats good idea by the way ! The only thing I have to care is that - the discovered IP should not be in the modified subnet range :)

Like that ! Lets me implement that way !

File: tests/foreman/api/test_discoveredhost.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py]: @rplevka Yeah ! Good idea ! Looks like the ipaddress module is inbuilt python lib.

File: tests/foreman/api/test_discoveredhost.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rplevka) [reply tests/foreman/api/test_discoveredhost.py:411]: yeah, good point. you can minimalize the chances by making the new range only 1-host wide + take it from the end of the old range.

[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py:411]: @rplevka Yeah ! Thats the plan !

File: tests/foreman/api/test_discoveredhost.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py]: Not using this module, doesnt need to make it that complex. Just using `discovered_host + 2` way to make it quite easier.

File: tests/foreman/api/test_discoveredhost.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py:411]: Done !

**Comment by jyejare** [comment]
@rplevka Revisit , I have encountered all your comments :)

**Review by rplevka — APPROVED** [review]

**Comment by jyejare** [comment]
@SatelliteQE/tier-2-reviewers Ping. Long pending PR, please review !

**Pull request merged**